python/demo/demo_mixed-poisson.py

```diff
@@ -109,88 +104,296 @@
-(tau, v) = TestFunctions(V)
+# The source function is set to be $f = 10\exp(-((x_{0} - 0.5)^2 +
+# (x_{1} - 0.5)^2) / 0.02)$:
 
+# +
 x = SpatialCoordinate(msh)
-f = 10.0 * exp(-((x[0] - 0.5) * (x[0] - 0.5) + (x[1] - 0.5) * (x[1] - 0.5)) / 0.02)
+f = 10 * exp(-((x[0] - 0.5) * (x[0] - 0.5) + (x[1] - 0.5) * (x[1] - 0.5)) / 0.02)
+# -
+
+# We now declare the blocked bilinear and linear forms. The rows of `a`
+# and `L` correspond to the $\tau$ and $v$ test functions, respectively.
+# The columns of `a` correspond to the $\sigma$ and $u$ trial functions,
+# respectively. Note that `a[1][1]` is empty, which is denoted by
+# `None`. This zero block is typical of a saddle-point problem. In the
+# `L[0]` block, the test function $\tau$ is multiplied by a zero
+# `Constant`, which is evaluated at runtime. We do this to preserve
+# knowledge of the test space in the block. *Note that the defined `L`
+# corresponds to $u_{0} = 0$ on $\Gamma_{D}$.*
 
+# +
 dx = Measure("dx", msh)
-a = inner(sigma, tau) * dx + inner(u, div(tau)) * dx + inner(div(sigma), v) * dx
-L = -inner(f, v) * dx
+a = [[inner(sigma, tau) * dx, inner(u, div(tau)) * dx], [inner(div(sigma), v) * dx, None]]
+L = [
+    inner(fem.Constant(msh, np.zeros(2, dtype=dtype)), tau) * dx,
+    -inner(f, v) * dx,
+]
+# -
+
+# We now compile the abstract/symbolic forms in `a` and `L` into
+# concrete instanced that can be assembled into matrix operators and
+# vectors, respectively.
 
-# Get subspace of V
-V0 = V.sub(0)
+# +
+a, L = fem.form(a, dtype=dtype), fem.form(L, dtype=dtype)
+# -
+
+# In preparation for Dirichlet boundary conditions, we use the function
+# `locate_entities_boundary` to locate mesh entities (facets) with which
+# degree-of-freedoms to be constrained are associated with, and then use
+# `locate_dofs_topological` to get the  degree-of-freedom indices. Below
+# we identify the degree-of-freedom in `V` on the (i) top ($x_{1} = 1$)
+# and (ii) bottom ($x_{1} = 0$) of the mesh/domain.
 
+# +
 fdim = msh.topology.dim - 1
-facets_top = mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 1.0))
-Q, _ = V0.collapse()
-dofs_top = fem.locate_dofs_topological((V0, Q), fdim, facets_top)
+dofs_top = fem.locate_dofs_topological(
+    V, fdim, mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 1.0))
+)
+dofs_bottom = fem.locate_dofs_topological(
+    V, fdim, mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 0.0))
+)
+# -
 
+# Now, we create Dirichlet boundary objects for the condition $\sigma
+# \cdot n = \sin(5 x_(0)$ on the top and bottom boundaries:
 
-def f1(x):
-    values = np.zeros((2, x.shape[1]))
-    values[1, :] = np.sin(5 * x[0])
-    return values
+# +
+f_h1 = fem.Function(V, dtype=dtype)
+f_h1.interpolate(lambda x: np.vstack((np.zeros_like(x[0]), np.sin(5 * x[0]))))
+bc_top = fem.dirichletbc(f_h1, dofs_top)
 
+f_h2 = fem.Function(V, dtype=dtype)
+f_h2.interpolate(lambda x: np.vstack((np.zeros_like(x[0]), -np.sin(5 * x[0]))))
```

Comment:
Could we instead use a single function `f`, where we highlight the possibility to restrict the interpolation to a sub-set of cells? 
This would highlight:
1. The usage of `top_cells = dolfinx.mesh.compute_incident_entities(msh.topology, top_facets, fdim, fdim+1)`
2. Interpolation into subset of cells:
    ```python
    f.interpolate(lambda x: np.stack((...), cells0=top_cells)
    f.interpolate(lambda x: np.stack((...), cells0=bottom_cells)
    f.x.scatter_forward()
    ```

Reply:
Ideally we would use a space that is defined only only BC boundary. How would this looks with the new mixed-domain support?

Reply:
It would look something like [this](https://github.com/jpdean/mixed_domain_demos/blob/main/neumann_bc.py). I can add it if you want?

Reply:
It would look something like [this](https://github.com/jpdean/mixed_domain_demos/blob/main/neumann_bc.py). I can add it if you want?

Reply:
> It would look something like [this](https://github.com/jpdean/mixed_domain_demos/blob/main/neumann_bc.py). I can add it if you want?

Yes, please do.

Reply:
Since these boundary conditions are enforced strongly, you have to use the space defined over the whole mesh (since you're setting the degrees of freedom values directly). We could add a non-zero natural boundary condition to demonstrate it though.

Reply:
I've got a generalized interpolation into normal based dof spaces at: https://fenicsproject.discourse.group/t/setting-scalar-function-for-bdm-boundary-condition-with-controlled-flux-mixed-poisson/13508/9?u=dokken
but it is currently quite verbose (the integration entity packing can now be removed as we have simplified this in DOLFINx).
However, i think we should proceed to add this to the C++ interface or as a Python convenience function.